Handle moving a UI-side composited window between displays with different properties

smfr opened the pull request:
#### a516b8527e73e0a16dfd0c660ca4ebd8d9b6130c
<pre>
Handle moving a UI-side composited window between displays with different properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=247932">https://bugs.webkit.org/show_bug.cgi?id=247932</a>
rdar://102346937

Reviewed by Tim Horton

On macOS windows can be moved between displays with different colorspaces, so with
UI-side compositing we need to ensure that we update layer backing stores with the
new colorspace.

To fix this, add the colorspace to RemoteLayerBackingStore::Parameters, and initialize
it in PlatformCALayerRemote::updateBackingStore() by getting the colorspace from the
RemoteLayerTreeContext (on macOS). This allows the existing call to `ensureBackingStore()`
to do the right thing when the colorspace changes.

We rely on the fact that moving between displays triggers repaints in all layers already
to make sure that we hit PlatformCALayerRemote::updateBackingStore().

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
(WebKit::RemoteLayerBackingStore::Parameters::operator== const):
(WebKit::RemoteLayerBackingStore::colorSpace const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::pixelFormat const):
(WebKit::RemoteLayerBackingStore::colorSpace const): Deleted.
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.cpp:
(WebKit::PlatformCALayerRemote::updateBackingStore):
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaMac::setColorSpace):

Canonical link: <a href="https://commits.webkit.org/258486@main">https://commits.webkit.org/258486@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/58c595ff3f62ed82d1b6b037d249eab328b690f0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102092 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11237 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35161 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111417 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12211 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2148 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94472 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109153 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107872 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9344 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92620 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37159 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/91222 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24098 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78884 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4794 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25525 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4903 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1967 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10960 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45021 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5819 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6651 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
```diff
@@ -60,6 +60,8 @@ class RemoteLayerTreeDrawingArea : public DrawingArea, public WebCore::GraphicsL
 protected:
     void updateRendering();
 
+    RemoteLayerTreeContext& remoteLayerTreeContext() const { return * m_remoteLayerTreeContext; }
```
`*m_remoteLayerTreeContext`

smfr committed:
Handle moving a UI-side composited window between displays with different properties
https://bugs.webkit.org/show_bug.cgi?id=247932
rdar://102346937

Reviewed by Tim Horton

On macOS windows can be moved between displays with different colorspaces, so with
UI-side compositing we need to ensure that we update layer backing stores with the
new colorspace.

To fix this, add the colorspace to RemoteLayerBackingStore::Parameters, and initialize
it in PlatformCALayerRemote::updateBackingStore() by getting the colorspace from the
RemoteLayerTreeContext (on macOS). This allows the existing call to `ensureBackingStore()`
to do the right thing when the colorspace changes.

We rely on the fact that moving between displays triggers repaints in all layers already
to make sure that we hit PlatformCALayerRemote::updateBackingStore().

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
(WebKit::RemoteLayerBackingStore::Parameters::operator== const):
(WebKit::RemoteLayerBackingStore::colorSpace const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::pixelFormat const):
(WebKit::RemoteLayerBackingStore::colorSpace const): Deleted.
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.cpp:
(WebKit::PlatformCALayerRemote::updateBackingStore):
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaMac::setColorSpace):

Canonical link: https://commits.webkit.org/258486@main
